### PR TITLE
New GraphQL filters

### PR DESCRIPTION
**Pull request by adriangmweb**
## Description
This new way of filtering is based on the Amplify one, which is way less limiting than the previous one.

## Changes
* The core `searcher` function has been changed to the new format. The old filter system is still compatible until we update the rest of the providers.
* The AWS `searcher-adapter` has been modified to translate these new filters to DynamoDB queries.
* Unit tests for AWS `searcher-adapter` has been added to ensure stability across all the different scenarios.

## Checks
- [x] Project Builds
- [x] Project passes tests and checks
- [x] Updated documentation accordingly - Just basic documentation. The rest will be added when the GraphQL schema gets updated in a new PR.
 
## Additional information
This is working just for retrieving and filtering ReadModels inside commands, events, etc. The GraphQL schema will be changed in a future PR.

### CODE REVIEW

File: packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts
Comment (NickSeagull) [inline packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts:70]: This is a great opportunity to extract this into a function. By the way, note the `console.log`, not sure if you meant to leave it there

File: packages/framework-provider-aws/src/library/searcher-adapter.ts
Comment (NickSeagull) [inline packages/framework-provider-aws/src/library/searcher-adapter.ts:13]: Can we use `unknown` instead of `any` here?

Comment (adriangmweb) [reply packages/framework-provider-aws/src/library/searcher-adapter.ts:13]: What would be the difference here @NickSeagull? I'm still learning some things about TS :)

Comment (NickSeagull) [reply packages/framework-provider-aws/src/library/searcher-adapter.ts:13]: The `unknown` type forces the programmer to cast the type to the appropriate type when its needed, while `any` can be passed around freely:

File: packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts
Comment (javiertoledo) [inline packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts:71]: After the previous changes you can also remove the `return true` at the end of the method.

File: packages/framework-provider-aws/src/library/searcher-adapter.ts
Comment (javiertoledo) [inline packages/framework-provider-aws/src/library/searcher-adapter.ts:158]: This switch is pretty similar to the one in `buildExpressionAttributeNames`. It makes me wonder if we could abstract this somehow to make both functions simpler 🤔 (Not a requirement for this PR though)

File: packages/framework-provider-aws/src/library/searcher-adapter.ts
Comment (javiertoledo) [inline packages/framework-provider-aws/src/library/searcher-adapter.ts:129]: What happens with this case? If we're planning to implement it in the future it could make sense to leave a TODO here:

```suggestion
      case 'includes':
        // TODO: To be implemented in the future. More details in issue #XYZ
        break
```

File: packages/framework-provider-aws/test/library/searcher-adapter.test.ts
Comment (javiertoledo) [inline packages/framework-provider-aws/test/library/searcher-adapter.test.ts:107]: Pretty cool DSL. Let's make sure it's properly documented before merging this PR, otherwise nobody will use this feature!

File: packages/framework-provider-local/src/library/searcher-adapter.ts
Comment (javiertoledo) [inline packages/framework-provider-local/src/library/searcher-adapter.ts:1]: Why renaming this type as `FilterOld`?

File: packages/framework-types/src/searcher.ts
Comment (javiertoledo) [inline packages/framework-types/src/searcher.ts:44]: Why deprecating it if we're changing the name anyway producing breaking changes? Would it make sense to remove this code entirely? What would be the consequences? Is other part of the system using it? Is this to avoid making the PR too long? If that's the case, make sure you add an issue to remove this deprecated code in the near future!

File: docs/README.md
Comment (javiertoledo) [inline docs/README.md:99]: It's weird that you changed the format in the whole TOC. Are you using the linter config in `.markdownlintrc`? If you are, there might be some extra options we should add to the config to make sure we all use the same.

**Review by javiertoledo — CHANGES_REQUESTED** [review]
Nice changes, just left a few comments to discuss. There are two points that I'd consider completing before merging:
1. Extensive documentation about the filtering DSL with all the available keywords and their meaning (We could agree on doing this in a different PR if we create an issue). 
2. We should also make sure that the DSL is the same for the API and for the in-handler interface (it's hard to see in the code review without digging much more in the code)
3. I'd consider removing the deprecated `filterOld` function and related code. As we're changing the function name, it's a breaking change anyway, and users could keep working with the current version until they make the changes for the new one...

File: packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts
Comment (alvaroloes) [inline packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts:7]: @adriangmweb This file should also be adapted to the new filter structure, but I see it keeps the same old structure, so subscriptions won't work properly.

_Maybe you want to do this in another PR? If so, please create the issue_

Context: When you subscribe to a ReadModel, you use the same filter as when querying them. We need to implement the behavior of those filters (as you did in AWS) but, this time, by filtering the array of read models in memory, in pure javascript (instead of with AWS SQL).

The GraphQL engine will call this code with an array of ReadModels of a specific type that are eligible to send to the user who subscribed, but we need to filter them using the filters the user specified in the Subscriptoin opreation.

File: packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts
Comment (alvaroloes) [reply packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts:71]: Take into account the previous comment marked with ⚠️ . The read model needs to pass _all the filters_

File: packages/framework-provider-aws/src/library/searcher-adapter.ts
Comment (alvaroloes) [reply packages/framework-provider-aws/src/library/searcher-adapter.ts:129]: I _think_ this is OK, because, after the include operation, there is only a value. There are no attribute names that needs to be accounted for in the AWS query.

Anyway, if that is the case, I agree that it would be great to have a comment explaining that.

File: packages/framework-provider-aws/src/library/searcher-adapter.ts
Comment (alvaroloes) [inline packages/framework-provider-aws/src/library/searcher-adapter.ts:39]: I see you call `resetAttrCount()` after `buildExpressionAttributeNames`, but that function doesn't use the global array that is reset. I guess you wanted to call it after `buildFilterExpression`.

Anyway, I think that using a global array (to this file) and knowing when I should reset it (it is not clear) is dangerous and can lead to bugs. Let me make you a proposal:

Instead of having a global array and resetting it after calling each of these two functions, _Why not creating two different empty arrays and passing them as parameters (this wouldn't be weird as the functions are recursive, so it is a common pattern to pass initial values as parameters)?_

The code would end up like this:
```suggestion
  const expresion = buildFilterExpression(filters, [])
  const names = buildExpressionAttributeNames(filters)
  const values = buildExpressionAttributeValues(filters, [])
```

This would erase possible subtle bugs when modifying this code and misplacing those "reset" calls or forget to add them, as it is not clear why we need to call them right here.

Check the other comments where I give hints of the other needed changes for this proposal

File: packages/framework-provider-aws/src/library/searcher-adapter.ts
Comment (alvaroloes) [inline packages/framework-provider-aws/src/library/searcher-adapter.ts:47]: **Related to the proposal to delete the global `attrCount` array:**

```suggestion
function buildFilterExpression(filters: FilterFor<any>, usedPlaceholders: Array<string>): string {
```

Also, this parameter should be passed to the different recursive calls we have here

File: packages/framework-provider-aws/src/library/searcher-adapter.ts
Comment (alvaroloes) [inline packages/framework-provider-aws/src/library/searcher-adapter.ts:59]: **Related to the proposal to delete the global `attrCount` array:**

```suggestion
          return buildOperation(propName, filter, usedPlaceholders)
```

File: packages/framework-provider-aws/src/library/searcher-adapter.ts
Comment (alvaroloes) [inline packages/framework-provider-aws/src/library/searcher-adapter.ts:66]: **Related to the proposal to delete the global `attrCount` array:**
```suggestion
  const holder = placeholderBuilderFor(propName, usedPlaceholders)
```

File: packages/framework-provider-aws/src/library/searcher-adapter.ts
Comment (alvaroloes) [inline packages/framework-provider-aws/src/library/searcher-adapter.ts:105]: **Related to the proposal to delete the global `attrCount` array:**

This code would not be needed anymore

File: packages/framework-provider-aws/src/library/searcher-adapter.ts
Comment (alvaroloes) [inline packages/framework-provider-aws/src/library/searcher-adapter.ts:113]: **Related to the proposal to delete the global `attrCount` array:**

Just using the parameter, instead of the global array:

```suggestion
function placeholderBuilderFor(propName: string): (valueIndex: number, valueSubIndex?: number, usedPlaceholders: Array<string>) => string {
  return (valueIndex: number, valueSubIndex?: number) => {
    const placeholder = `:${propName}_${valueIndex}` + (typeof valueSubIndex === 'number' ? `_${valueSubIndex}` : '')
    if (usedPlaceholders.includes(placeholder)) return placeholderBuilderFor(propName, undefined, usedPlaceholders)(valueIndex + 1)
    usedPlaceholders.push(placeholder)
    return placeholder
  }
}
```

File: packages/framework-provider-aws/src/library/searcher-adapter.ts
Comment (alvaroloes) [inline packages/framework-provider-aws/src/library/searcher-adapter.ts:143]: Do the same changes that we did in "buildFilterExpression" regarding removing the attrCoung global array

File: packages/framework-provider-aws/test/library/searcher-adapter.test.ts
Comment (alvaroloes) [reply packages/framework-provider-aws/test/library/searcher-adapter.test.ts:107]: 1000% agree. Now is the time to document this thoroughly

File: packages/framework-provider-aws/test/library/searcher-adapter.test.ts
Comment (alvaroloes) [inline packages/framework-provider-aws/test/library/searcher-adapter.test.ts:202]: LOVE THE TESTS!

File: packages/framework-provider-local/src/library/searcher-adapter.ts
Comment (alvaroloes) [reply packages/framework-provider-local/src/library/searcher-adapter.ts:1]: I think he did that to avoid the need of changing all the providers at once to the new filters. So he left the old filters in place and used only the new ones in AWS

File: packages/framework-types/src/envelope.ts
Comment (alvaroloes) [inline packages/framework-types/src/envelope.ts:46]: I'm not sure if it won't give you problems, but Could you please try to change this to `FilterFor<unknown>`? This way we avoid `any` which is a bit dangerous in Typescript.

If you find too many problems after the change, forget it

**Review by alvaroloes — COMMENTED** [review]
Left some commange @adriangmweb Thanks a lot for this!!

File: packages/framework-provider-local/src/library/searcher-adapter.ts
Comment (javiertoledo) [reply packages/framework-provider-local/src/library/searcher-adapter.ts:1]: 🤔 and that's why we need to move the core features to separate rockets too...

File: packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts
Comment (javiertoledo) [reply packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts:41]: Oh, I didn't notice the `for` at the beginning of the function. Still, you could extract the switch to a new function and apply the suggested changes to make this code a bit more readable. Anyway, this was just a suggestion, @adriangmweb decides!

File: docs/README.md
Comment (adriangmweb) [reply docs/README.md:1604]: Completely agree with this!

File: packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts
Comment (adriangmweb) [reply packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts:7]: This won't actually be able to test until we got the GraphQL new filter working, right? I'll create the issue to solve it along with the new schema

File: packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts
Comment (adriangmweb) [reply packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts:41]: I'll be working on refactoring this file once I start working on the new GraphQL schema so it can be properly tested. I created an issue for this https://github.com/boostercloud/booster/issues/600

File: packages/framework-provider-aws/src/library/searcher-adapter.ts
Comment (adriangmweb) [reply packages/framework-provider-aws/src/library/searcher-adapter.ts:129]: This is because when there is an includes operation it shouldn't enter in the `default` case but get ingnored instead.

File: packages/framework-provider-aws/test/library/searcher-adapter.test.ts
Comment (adriangmweb) [reply packages/framework-provider-aws/test/library/searcher-adapter.test.ts:107]: I think it would make more sense to document it on the GraphQL new schema branch.

File: docs/README.md
Comment (adriangmweb) [reply docs/README.md:99]: I haven't change anything but if I'm the only one experiencing this, I'll love to compare our rules to find what I'm missing 😅

File: packages/framework-provider-aws/src/library/searcher-adapter.ts
Comment (adriangmweb) [reply packages/framework-provider-aws/src/library/searcher-adapter.ts:39]: This is awesome @alvaroloes! 💯  I just implemented it and even improve it a little, I'll show you shortly!

File: packages/framework-provider-aws/src/library/searcher-adapter.ts
Comment (javiertoledo) [reply packages/framework-provider-aws/src/library/searcher-adapter.ts:129]: Then I'd add a comment explaining that or someone could end up removing that code thinking that's useless..

File: packages/framework-provider-aws/test/library/searcher-adapter.test.ts
Comment (javiertoledo) [reply packages/framework-provider-aws/test/library/searcher-adapter.test.ts:107]: Well, you have to document two things: how you use it from a handler and how you'll use it when it's available from the API. I'd add the docs on each PR to make sure it's not forgotten later

Comment (adriangmweb) [reply packages/framework-provider-aws/test/library/searcher-adapter.test.ts:107]: Wouldn't be enough with the current `Getting and filtering read models data at code level` section I added on the docs?

File: packages/framework-types/src/searcher.ts
Comment (adriangmweb) [reply packages/framework-types/src/searcher.ts:44]: Because it's still used in the rest of the providers, so it's renamed in order to be able to make them know it's an old and deprecated functionality. I just created a new issue explaining that we should migrate all providers to the new system before removing it. https://github.com/boostercloud/booster/issues/601

File: packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts
Comment (adriangmweb) [reply packages/framework-core/src/services/pub-sub/read-model-pub-sub.ts:7]: Here's the issue https://github.com/boostercloud/booster/issues/602

**Review by javiertoledo — APPROVED** [review]
Approving, as all my concerns have been either addressed or moved to new issues. Please make sure that documentation is addressed in the short term, or this feature won't be used.

**Review by davidverdu — APPROVED** [review]
LGTM, provided that some documentation is added in the future. These are amazing changes! (nonetheless, there are things that appear to be complex for me at first glance)

**Review by alvaroloes — COMMENTED** [review]
I have now realized that I had this review pending to submit from the previous week 🤦

**Pull request merged**
